akka-stream-tests/src/test/scala/akka/stream/scaladsl/BoundedSourceQueueSpec.scala

```diff
@@ -132,29 +132,29 @@ class BoundedSourceQueueSpec extends StreamSpec("""akka.loglevel = debug
           def runLoop(): Unit = {
             val r = ThreadLocalRandom.current()
 
-            while (true) {
+            var done = false
+            while (!done) {
               val i = r.nextInt(0, Int.MaxValue)
               queue.offer(i) match {
                 case QueueOfferResult.Enqueued =>
                   counter.addAndGet(i)
                   numElemsEnqueued += 1
                 case QueueOfferResult.Dropped =>
                   numElemsDropped += 1
-                case _: QueueCompletionResult => return // other thread completed
+                case unexpected => throw new IllegalStateException(s"Saw $unexpected should not happen in this test")
               }
 
               if ((i % stopProb) == 0) { // probabilistic exit condition
-                queue.complete()
-                return
-              }
-
-              if (i % 100 == 0) Thread.sleep(1) // probabilistic producer throttling delay
+                done = true
+              } else if (i % 100 == 0) Thread.sleep(1) // probabilistic producer throttling delay
             }
           }
 
-          barrier.countDown()
-          barrier.await() // wait for all threads being in this state before starting race
+          startBarrier.countDown()
+          startBarrier.await() // wait for all threads being in this state before starting race
           runLoop()
+          stopBarrier.countDown()
+          stopBarrier.await()
```

Comment:
If I read it correctly you can probably leave only the `await` at line 168 that actually await for all of the Threads.

Reply:
True. I'll remove it.